File: facilitator-solutions/javascript/rung-6-on-byte-array.js
// Rung 6 applied to rung 4, WITHOUT the bitset. Rungs 1, 2, 3, 4 and 6.
//
// A new Uint8Array is already zeroed. Invert the sense so that zero means prime and
// the fill disappears entirely, every lap - no bitset required underneath it.
//
// This matters more in JavaScript than anywhere else, because rung 5 measures as a
// regression here: the odds only Uint8Array is JavaScript's best result on the
// ladder, and this is the cheapest thing that can be added to it.
//
// No tail word problem. The array is exactly count long, so there are no spare
// slots past n to read back as prime.

export const getSieve = (sieveSize) => {
	const count = (sieveSize + 1) >> 1;

	// Already zeroed, and 0 now means prime - so no fill at all.
	const sieve = new Uint8Array(count);

	const limit = Math.floor(Math.sqrt(sieveSize));

	return {
		run: () => {
			for (let factor = 3; factor <= limit; factor += 2) {
				if (sieve[factor >> 1] === 0) {
					for (let q = (factor * factor) >> 1; q < count; q += factor) {
						sieve[q] = 1;
					}
				}
			}
		},
		results: () => {
			const primes = [];
			if (sieveSize >= 2) {
				primes.push(2);
			}
			for (let i = 1; i < count; i++) {
				if (sieve[i] === 0) {
					primes.push(2 * i + 1);
				}
			}
			return primes;
		},
	};
};

File: facilitator-solutions/go/rung-6-on-byte-array.go
package main

import "math"

// Rung 6 applied to rung 4, WITHOUT the bitset. Rungs 2, 3, 4 and 6.
//
// make([]bool, n) is already zeroed, and the starting point then loops over the
// whole thing writing true. Invert the sense so false means prime and that entire
// pass disappears - without needing the bitset underneath it.
//
// Measured because the nudge order depends on it: rung 5 can measure negative,
// this cannot, because it only removes work.
//
// No tail word problem. The slice is exactly count long.

type primeSieve struct {
	flags []bool
	size  int
}

func getSieve(sieveSize int) primeSieve {
	// make() already zeroes this, and false now means prime - so no fill at all.
	count := (sieveSize + 1) / 2
	return primeSieve{flags: make([]bool, count), size: sieveSize}
}

func (sieve primeSieve) run() {
	limit := int(math.Sqrt(float64(sieve.size)))
	count := len(sieve.flags)

	for factor := 3; factor <= limit; factor += 2 {
		if !sieve.flags[factor>>1] {
			for q := (factor * factor) >> 1; q < count; q += factor {
				sieve.flags[q] = true
			}
		}
	}
}

func (sieve primeSieve) result() []int {
	var primes []int

	if sieve.size >= 2 {
		primes = append(primes, 2)
	}

	for i := 1; i < len(sieve.flags); i++ {
		if !sieve.flags[i] {
			primes = append(primes, 2*i+1)
		}
	}

	return primes
}
